garage/algos/ddqn.py

```diff
@@ -0,0 +1,339 @@
+        gamma : float
+            Discount factor in the Q-value update. This value has to be in the
+            range of [0.0, 1.0]. A value of 0 makes the agent only consider
+            immediate rewards, while a value of 1 makes the agent only consider
+            later rewards.
+        alpha : float
+            Learning_rate in the Q-value update. This value has to be in the
+            range of [0.0, 1.0]. Exploitation of prior knowledge is used when
+            alpha is zero, and exploration when alpha is one.
+        memory_bank_size : int
+            Maximum number of transitions that will be saved in the memory
+            bank. Oldest transitions will be discarded once memory is full.
+            If set as -1, memory is not limited.
+        units_per_layer : tuple of ints
+            Number of units for each layer in the neural network.
+        weights_per_layer : tuple of weights
+            Initial value of weights for each layer in the neural network.
+        bias_per_layer : tuple of biases
+            Initial value of biases for each layer in the neural network.
+        nonlinearity_per_layer : tuple of non-linearities
+            Non-linearities applied to the activations of each layer in the
+            neural network.
+
+        """
+        assert episodes > 0
+        assert tau > 0
+        assert batch_size > 0
+        assert exploration_steps >= batch_size
+        assert (1.0 >= gamma and gamma >= 0.0)
+        assert (1.0 >= alpha and alpha >= 0.0)
+        assert ((len(units_per_layer) == len(weights_per_layer)) and
+                (len(weights_per_layer) == len(bias_per_layer)) and
+                (len(bias_per_layer) ==
+                 len(nonlinearity_per_layer)))  # yapf: disable
+        self._env = env
+        self._exploration_strategy = exploration_strategy
+        self._episodes = episodes
+        self._tau = tau
+        self._batch_size = batch_size
+        self._exploration_steps = exploration_steps
+        self._gamma = gamma
+        self._alpha = alpha
+        self._units = units_per_layer
+        self._weights = weights_per_layer
+        self._bias = bias_per_layer
+        self._nonlinearities = nonlinearity_per_layer
+
+        self._step_total = 0
+        self._render = False
+        self._continue_trainig = True
+        # Observers of the training
+        self._training_obs = []
+
+        self._memory = MemoryBank()
+        # Placeholders for variables used in the neural networks
+        self._obs = tensor.matrix("Observation")
+        self._obs_next = tensor.matrix("ObservationNext")
+        self._action = tensor.ivector("Action")
+        self._reward = tensor.vector("Reward")
+        self._done = tensor.vector("Done")
+        # Build the networks
+        self._target_nn = self._create_network()
+        self._online_nn = self._create_network()
+        self._build_graph()
+
+    def _create_network(self):
+        """Create a neural network.
```

Comment:
I think this should be moved to something like q_function. Please remember to create some reusable components.